python/dgl/nn/pytorch/explain/gnnexplainer.py

```diff
@@ -425,3 +398,395 @@ def explain_graph(self, graph, feat, **kwargs):
+                feat_masks[node_type] = nn.Parameter(torch.randn(1, feat_size, device=device) * std)
+
+        edge_masks = {}
+        for canonical_etype in graph.canonical_etypes:
+            src = canonical_etype[0]
+            num_nodes = graph.number_of_nodes(src)
+            num_edges = graph.number_of_edges(canonical_etype)
+            std = nn.init.calculate_gain('relu')
+            if num_nodes > 0:
+                std *= sqrt((2.0 / (2 * num_nodes)))
+            edge_masks[canonical_etype] = nn.Parameter(torch.randn(num_edges, device=device) * std)
+
+        return feat_masks, edge_masks
+
+    def _loss_regularize(self, loss, feat_masks, edge_masks):
+        r"""Add regularization terms to the loss.
+        Parameters
+        ----------
+        loss : Tensor
+            Loss value.
+        feat_mask : Dictionary of Tensor
+            Dictionary of { ntypes: features }
+            Feature mask of shape :math:`(1, D)`, where :math:`D`
+            is the feature size.
+        edge_mask : Dictionary of Tensor
+            Dictionary of { canonical_etypes: features }
+            Edge mask of shape :math:`(E)`, where :math:`E`
+            is the number of edges.
+        Returns
+        -------
+        Tensor
+            Loss value with regularization terms added.
+        """
+        # epsilon for numerical stability
+        eps = 1e-15
+
+        for edge_mask in edge_masks.values():
+            edge_mask = edge_mask.sigmoid()
+            # Edge mask sparsity regularization
+            loss = loss + self.alpha1 * torch.sum(edge_mask)
+            # Edge mask entropy regularization
+            ent = - edge_mask * torch.log(edge_mask + eps) - \
+                (1 - edge_mask) * torch.log(1 - edge_mask + eps)
+            loss = loss + self.alpha2 * ent.mean()
+
+        for feat_mask in feat_masks.values():
+            feat_mask = feat_mask.sigmoid()
+            # Feature mask sparsity regularization
+            loss = loss + self.beta1 * torch.mean(feat_mask)
+            # Feature mask entropy regularization
+            ent = -feat_mask * torch.log(feat_mask + eps) - \
+                (1 - feat_mask) * torch.log(1 - feat_mask + eps)
+            loss = loss + self.beta2 * ent.mean()
+
+        return loss
+
+    def explain_node(self, ntype, node_id, graph, feat, **kwargs):
+        r"""Learn and return a node feature mask and subgraph that play a
+        crucial role to explain the prediction made by the GNN for node
+        :attr:`node_id`.
+        Parameters
+        ----------
+        ntype: str
+            The node type of the node to explain.
+        node_id : int
+            The node to explain.
+        graph : DGLHeteroGraph
+            A heterogeneous graph.
+        feat : Dictionary of Tensor
+            Dictionary of { ntypes: features }
+            The input feature of shape :math:`(N, D)`. :math:`N` is the
+            number of nodes, and :math:`D` is the feature size.
```

Comment:
Ditto

Reply:
made changes in the recent commit, can you please take a look and let me know if further improvement is needed in the explanation?